File: utils/dump_page_tables.py
#!/usr/bin/python
# Takes a VirtualBox memory dump (as an .elf file), and given the offset
# of the memory dump, the cr3 value and the location of the new page
# tables, will dump all of them from the memory dump.
#
# These items can be found like so:
# Start VirtualBox to be able to create a memory dump:
#   VirtualBox --dbg --startvm nebulae
# Create the memory dump at the appropriate time:
#   VBoxManage debugvm "nebulae" dumpvmcore --filename nebdump.elf
# Find the offset of the elf64 section that contains the memory dump: 
#   objdump -h nebdump.elf | egrep -w "(Idx|load1)"
# CR3 and the new page table location is provided by nebulae and/or
# Vbox.log, depending on what you're doing.  You could also just dump
# it to screen before panicing

import sys
import array as arr
from pathlib import Path

if len(sys.argv) != 4:
    print(f"Usage: dump_page_tables.py <filename.elf> <memory_start_offset> <cr3>")
    sys.exit()

elf_file = Path(sys.argv[1])

if not elf_file.is_file():
    print(f"{sys.argv[1]} does not exist!")
    sys.exit()

memory_offset = int(sys.argv[2])

if memory_offset == 0:
    print(f"Please enter a valid memory start offset!")
    sys.exit()

cr3 = int(sys.argv[3])

if cr3 == 0:
    print(f"Please enter a valid value for the CR3 register!")
    sys.exit()

pml4_table = arr.array('L')

x64_4KB_ALIGNMENT = 0x000FFFFFFFFFF000

# Validate input file
try:
    with elf_file.open("rb") as elf:
        # at least verify this is an elf file
        elf_sig = elf.read(4)
        if elf_sig != b"\x7fELF":
            print(f"ELF signature not found!")
            sys.exit()

        # dump the pml4 table
        for pml4_i in range(512):
            # seek to the proper pml4 entry
            elf.seek(memory_offset + (cr3 + (pml4_i * 8)))
            pml4_table.append(int.from_bytes(elf.read(8), byteorder='little'))
            print(f"PML4[{pml4_i}] == " + hex(pml4_table[pml4_i]))

            # if this pml4 table is empty, we're done
            if pml4_table[pml4_i] == 0:
                continue

            # each pdpt has 512 entries
            pdpt = arr.array('L')
            for pdpt_i in range(512):
                # seek to the pdpte this pml4 entry points to
                elf.seek(memory_offset + (pml4_table[pml4_i] & x64_4KB_ALIGNMENT) + (pdpt_i * 8))
                
                pdpt.append(int.from_bytes(elf.read(8), byteorder='little'))
                print(f"PDPT[{pdpt_i}] == " + hex(pdpt[pdpt_i]))

                # see if this points to a 1GB page
                if (pdpt[pdpt_i] & 0x80) == 0x80:
                    print(f"1GB Page Table Found @ {pdpt_i} == " + hex(pdpt[pdpt_i]))
                    continue
                
                # if this pdpt is empty, we're done
                if pdpt[pdpt_i] == 0:
                    continue

                # each pd has 512 entries
                pd = arr.array("L")
                for pd_i in range(512):
                    # seek to the page directory this pdpt entry points to
                    elf.seek(memory_offset + (pdpt[pdpt_i] & x64_4KB_ALIGNMENT) + (pd_i * 8))    
                    
                    pd.append(int.from_bytes(elf.read(8), byteorder='little'))
                    print(f"PD[{pd_i}] == " + hex(pd[pd_i]))

                    # see if this pde points to a page table
                    if (pd[pd_i] & 0x80) == 0x80:
                        print(f"2MB Page Table Found @ {pd_i} == " + hex(pd[pd_i]))
                        continue

                    # jump to the page table this page directory entry points to
                    elf.seek(memory_offset + (pd[pd_i] & x64_4KB_ALIGNMENT))

                    # print each of the page table entries
                    pt = arr.array("L")
                    for pt_i in range(512):
                        pt.append(int.from_bytes(elf.read(8), byteorder='little'))
                        print(f"4KB Page Table Found @ {pt_i} == " + hex(pt[pt_i]))

except Exception as inst:
    print(type(inst))       # the exception instance
    print(inst.args)        # arguments stored in .args
    print(inst)             # __str__ allows args to be printed directly,
                            # but may be overridden in exception subclasses
    sys.exit()

sys.exit()

File: utils/dump_page_tables.php
<?php
// Copyright (c) 2018-2019 Nebulae Foundation. All rights reserved.
//
// Redistribution and use in source and binary forms, with or without 
// modification, are permitted provided that the following conditions are met:
// 
// 1. Redistributions of source code must retain the above copyright notice, 
// this list of conditions and the following disclaimer.
// 
// 2. Redistributions in binary form must reproduce the above copyright notice, 
// this list of conditions and the following disclaimer in the documentation 
// and/or other materials provided with the distribution.
// 
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" 
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE 
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
// POSSIBILITY OF SUCH DAMAGE.

// Takes a VirtualBox memory dump (as an .elf file), and given the offset
// of the memory dump, the cr3 value and the location of the new page
// tables, will dump all of them from the memory dump.
//
// These items can be found like so:
// Start VirtualBox to be able to create a memory dump:
//   VirtualBox --dbg --startvm nebulae
// Create the memory dump at the appropriate time:
//   VBoxManage debugvm "nebulae" dumpvmcore --filename nebdump.elf
// Find the offset of the elf64 section that contains the memory dump: 
//   objdump -h nebdump.elf | egrep -w "(Idx|load1)"
// CR3 and the new page table location is provided by nebulae and/or
// Vbox.log, depending on what you're doing.  You could also just dump
// it to screen before panicing

require_once("uint64.php");

if ($argc != 5) {
    fwrite(STDOUT, "USAGE: {$argv[0]} <dumpfile> <memory_section_offset> <cr3> <new_page_table_location>\n\n");
    exit();
}

if (!file_exists($argv[1])) {
    fwrite(STDOUT, "File {$argv[1]} not found\n");
    exit();
}

if (!is_numeric($argv[2]) || !is_numeric($argv[3]) || !is_numeric($argv[4])) {
    fwrite(STDOUT, "Please specify integer arguments for memory_section_offset, cr3, and new_page_table_location\n");
    exit();
}

if (is_null(($elf = fopen($argv[1], "r")))) {
    fwrite(STDOUT, "Unable to open file: {$argv[1]}\n");
    exit();
}

// Assign the start of the physical memory dump
// within the elf file
$memory_start_offset = intval($argv[2]);

// 
SeekToMemStart();


// This function sets the file pointer back to the 
// offset of the start of the physical memory dump 
function SeekToMemStart() {
    global $memory_start_offset, $elf;

    fseek($elf, $memory_start_offset, SEEK_SET);
}

?>
